src/main/java/org/apache/cassandra/sidecar/concurrent/TaskExecutorPool.java

```diff
@@ -0,0 +1,277 @@
+     *
+     * @param id The id of the timer to cancel
+     * @return {@code true} if the timer was successfully cancelled, {@code false} otherwise
+     */
+    public boolean cancelTimer(long id)
+    {
+        return vertx.cancelTimer(id);
+    }
+
+    @Override
+    public <T> void executeBlocking(Handler<Promise<T>> blockingCodeHandler,
+                                    boolean ordered,
+                                    Handler<AsyncResult<T>> asyncResultHandler)
+    {
+        workerExecutor.executeBlocking(blockingCodeHandler, ordered, asyncResultHandler);
+    }
+
+    @Override
+    public <T> Future<T> executeBlocking(Handler<Promise<T>> blockingCodeHandler,
+                                         boolean ordered)
+    {
+        return Timer.measureTimeTaken(workerExecutor.executeBlocking(blockingCodeHandler, ordered), this::recordTimeTaken);
+    }
+
+    @Override
+    public <T> Future<T> executeBlocking(Callable<T> blockingCodeHandler, boolean ordered)
+    {
+        return Timer.measureTimeTaken(workerExecutor.executeBlocking(blockingCodeHandler, ordered), this::recordTimeTaken);
+    }
+
+    @Override
+    public <T> void executeBlocking(Handler<Promise<T>> blockingCodeHandler,
+                                    Handler<AsyncResult<T>> asyncResultHandler)
+    {
+        workerExecutor.executeBlocking(blockingCodeHandler, asyncResultHandler);
+    }
+
+    @Override
+    public <T> Future<T> executeBlocking(Handler<Promise<T>> blockingCodeHandler)
+    {
+        return Timer.measureTimeTaken(workerExecutor.executeBlocking(blockingCodeHandler), this::recordTimeTaken);
+    }
+
+    /**
+     * Records time taken for tasks executed by {@link TaskExecutorPool}
+     * @param duration time taken by a task
+     */
+    protected abstract void recordTimeTaken(long duration);
+
+    @Override
+    public void close(Handler<AsyncResult<Void>> handler)
+    {
+        throw new UnsupportedOperationException("Closing TaskExecutorPool is not supported!");
+    }
+
+    @Override
+    public Future<Void> close()
+    {
+        throw new UnsupportedOperationException("Closing TaskExecutorPool is not supported!");
+    }
+
+    Future<Void> closeInternal()
+    {
+        return workerExecutor == null
+               ? Future.succeededFuture()
+               : workerExecutor.close();
+    }
+
+    /**
+     * {@link ServiceTaskExecutorPool} is used for executing tasks that are short lived and not expected to block for
+     * too long, therefore will free up resources more quickly
+     */
+    static class ServiceTaskExecutorPool extends TaskExecutorPool
+    {
+        private final ResourceMetrics metrics;
+
+        ServiceTaskExecutorPool(Vertx vertx, WorkerPoolConfiguration config, ResourceMetrics metrics)
+        {
+            super(vertx, config);
+            this.metrics = metrics;
+        }
+
+        @Override
+        protected void recordTimeTaken(long duration)
+        {
+            if (metrics == null)
```

Comment:
will `metrics` ever be null? it doesn't seem like we do null checks for other metrics anywhere else in the code

Reply:
This one can be, because we might want to allow creation of `ExecutorPools` without metrics. Since `ExecutorPools` is also created from integration tests. Thats why we do the null check only here.